src/lib/comms/sol-http-common.c

```diff
@@ -74,52 +130,645 @@ sol_http_param_free(struct sol_http_param *params)
+        SOL_INT_CHECK_GOTO(r, < 0, err_exit);
+
+        param.value.key_value.key = sol_buffer_get_slice(&decoded_key);
+        param.value.key_value.value = sol_buffer_get_slice(&decoded_value);
+        SOL_DBG("Query key: %.*s Query value: %.*s",
+            SOL_STR_SLICE_PRINT(param.value.key_value.key),
+            SOL_STR_SLICE_PRINT(param.value.key_value.value));
+
+        param.type = type;
+
+        if (!sol_http_param_add_copy(params, param)) {
+            SOL_WRN("Could not alloc the param %.*s : %.*s",
+                SOL_STR_SLICE_PRINT(param.value.key_value.key),
+                SOL_STR_SLICE_PRINT(param.value.key_value.value));
+            r = -ENOMEM;
+            goto err_exit;
+        }
+
+        sol_buffer_fini(&decoded_key);
+        sol_buffer_fini(&decoded_value);
+    }
+
+    sol_vector_clear(&tokens);
+    return r;
+
+err_exit:
+    sol_buffer_fini(&decoded_key);
+    sol_buffer_fini(&decoded_value);
+    sol_vector_clear(&tokens);
+    return r;
 }
 
 SOL_API int
-sol_http_escape_string(char **escaped, const char *value)
+sol_http_create_uri(char **url_out, const struct sol_http_url url,
+    const struct sol_http_param *params)
 {
+    struct sol_str_slice scheme, sep;
+    struct sol_buffer buf;
     int r;
-    size_t value_len;
-    struct sol_buffer buf = SOL_BUFFER_INIT_EMPTY;
-
-    SOL_NULL_CHECK(value, -EINVAL);
-    SOL_NULL_CHECK(escaped, -EINVAL);
-
-    value_len = strlen(value);
-
-    while (value_len--) {
-        unsigned char c = *value;
-        switch (c) {
-        case '0': case '1': case '2': case '3': case '4':
-        case '5': case '6': case '7': case '8': case '9':
-        case 'a': case 'b': case 'c': case 'd': case 'e':
-        case 'f': case 'g': case 'h': case 'i': case 'j':
-        case 'k': case 'l': case 'm': case 'n': case 'o':
-        case 'p': case 'q': case 'r': case 's': case 't':
-        case 'u': case 'v': case 'w': case 'x': case 'y': case 'z':
-        case 'A': case 'B': case 'C': case 'D': case 'E':
-        case 'F': case 'G': case 'H': case 'I': case 'J':
-        case 'K': case 'L': case 'M': case 'N': case 'O':
-        case 'P': case 'Q': case 'R': case 'S': case 'T':
-        case 'U': case 'V': case 'W': case 'X': case 'Y': case 'Z':
-        case '_': case '~': case '.': case '-':
-            r = sol_buffer_append_char(&buf, c);
-            SOL_INT_CHECK_GOTO(r, < 0, end);
-            break;
-        default:
-            r = sol_buffer_append_printf(&buf, "%%%02X", c);
-            SOL_INT_CHECK_GOTO(r, < 0, end);
+
+    SOL_NULL_CHECK(url_out, -EINVAL);
+
+    sol_buffer_init(&buf);
+
+    if (url.scheme.len)
+        scheme = url.scheme;
+    else
+        scheme = sol_str_slice_from_str("http");
+
+    r = sol_buffer_append_slice(&buf, scheme);
+    SOL_INT_CHECK_GOTO(r, < 0, exit);
+
+    if (!sol_str_slice_str_caseeq(scheme, "mailto") &&
+        !sol_str_slice_str_caseeq(scheme, "news"))
+        sep = sol_str_slice_from_str("://");
```

Comment:
this is not right, there should be no special case... and mailto AFAIK has no "://".

Even "news", it requires a server to get the "//", so just do:

```c
    r = sol_buffer_append_slice(&buf, scheme);
    ...
    r = sol_buffer_append_char(&buf, ':');

    if (url.host.len) {
       r = sol_buffer_append_slice(&buf, SOL_STR_SLICE("//"));
       ...
       if (url.user.len) {
           r = sol_buffer_append_slice(&buf, url.user); // we may have to encode this? plz check
           if (url.password.len) {
               r = sol_buffer_append_char(&buf, ':');
               r = sol_buffer_append_slice(&buf, url.password); // we may have to encode this? plz check
         }
          r = sol_buffer_append_char(&buf, '@');
    }
    r = sol_buffer_append_slice(&buf, url.host);
    if (url.port) {
       r = sol_buffer_append_char(&buf, ':');
       r = sol_buffer_append_printf(&buf, "%u", url.port);
   }
```